### ISSUE TITLE
NoSuchElementException on http request timeout

### ISSUE THREAD

**Issue by sschaef** [akka/akka#20388]
The following exception is thrown when a request to the server takes too long to complete:

```
backend [ERROR] [04/24/2016 13:53:36.994] [default-akka.actor.default-dispatcher-4] [akka.actor.ActorSystemImpl(default)] Internal server error, sending 500 response
backend java.util.NoSuchElementException: head on empty queue
backend         at scala.collection.immutable.Queue.head(Queue.scala:82)
backend         at akka.http.impl.engine.server.HttpServerBluePrint$RequestTimeoutSupport$$anon$6$$anon$7.emitTimeoutResponse(HttpServerBluePrint.scala:279)
backend         at akka.http.impl.engine.server.HttpServerBluePrint$RequestTimeoutSupport$$anon$6$$anon$7$$anonfun$7.apply(HttpServerBluePrint.scala:272)
backend         at akka.http.impl.engine.server.HttpServerBluePrint$RequestTimeoutSupport$$anon$6$$anon$7$$anonfun$7.apply(HttpServerBluePrint.scala:272)
backend         at akka.stream.impl.fusing.GraphInterpreter.runAsyncInput(GraphInterpreter.scala:569)
backend         at akka.stream.impl.fusing.GraphInterpreterShell.receive(ActorGraphInterpreter.scala:417)
backend         at akka.stream.impl.fusing.ActorGraphInterpreter.akka$stream$impl$fusing$ActorGraphInterpreter$$processEvent(ActorGraphInterpreter.scala:593)
backend         at akka.stream.impl.fusing.ActorGraphInterpreter$$anonfun$receive$1.applyOrElse(ActorGraphInterpreter.scala:604)
backend         at akka.actor.Actor$class.aroundReceive(Actor.scala:482)
backend         at akka.stream.impl.fusing.ActorGraphInterpreter.aroundReceive(ActorGraphInterpreter.scala:519)
backend         at akka.actor.ActorCell.receiveMessage(ActorCell.scala:526)
backend         at akka.actor.ActorCell.invoke(ActorCell.scala:495)
backend         at akka.dispatch.Mailbox.processMailbox(Mailbox.scala:257)
backend         at akka.dispatch.Mailbox.run(Mailbox.scala:224)
backend         at akka.dispatch.Mailbox.exec(Mailbox.scala:234)
backend         at scala.concurrent.forkjoin.ForkJoinTask.doExec(ForkJoinTask.java:260)
backend         at scala.concurrent.forkjoin.ForkJoinPool$WorkQueue.runTask(ForkJoinPool.java:1339)
backend         at scala.concurrent.forkjoin.ForkJoinPool.runWorker(ForkJoinPool.java:1979)
backend         at scala.concurrent.forkjoin.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:107)
```

The intended response is still sent to the client but this error message is printed in the log.

I use Akka 2.4.3

**Comment by ktoso** [comment]
Thanks for reporting

**Comment by rbudzko** [comment]
`openTimeouts.headOption.exists(_ eq response._1)` should fix this issue. Not sure how to verify it with tests without suspending and resuming threads.

The whole change would be:
````
// the application response might has already arrived after we scheduled the timeout response (which is close but ok)
// or current head (same reason) is not for response the timeout has been scheduled for
if (openTimeouts.headOption.exists(_ eq response._1)) {
   emit(responseOut, response._2, () ⇒ completeStage())
}
````

**Comment by ktoso** [comment]
Please submit a PR @rbudzko, let's see

**Comment by rbudzko** [comment]
#20458

**Comment by ktoso** [comment]
(ref the other way please @rbudzko ;-))

**Comment by rbudzko** [comment]
(how :-) ? by amending commit message?)

**Comment by ktoso** [comment]
(Nah, just "Resolves #20388" from the PR, instead refering to the PR from here - total detail, don't worry about it :))
